sympy/sets/sets.py

```diff
@@ -1465,12 +1465,32 @@ def reduce(args):
                 unk = [x for x in s
                        if any(other.contains(x) not in (True, False) for other in other_args)]
                 if unk:
+                    for val in other_args:
+                        if val.is_FiniteSet:
+                            # all values from val which are contained in `s`
+                            # and are non-numbers
+                            t = [x for x in s if val.contains(x) is true and not x.is_number]
```

Comment:
> It seems that all non-numbers are treated in the same way as (SymPy) symbols. Is there a reason why only numbers should be dealt with separately?

That was an error that i made. I have now used `.atoms(Symbol)` to remove that problem. I was in-fact leaving out big space of error. But now using `.atoms(Symbol)` has got me thinking that: Is this an expensive operation?. If it is expensive we could also use `*args` from user to simplify using this may be.